### PR TITLE
[JENKINS-55813] Improve analysis of the AD attributes

### PR DESCRIPTION
**Pull request by Wadeck**
See [JENKINS-55813](https://issues.jenkins-ci.org/browse/JENKINS-55813).

### CODE REVIEW

File: src/main/java/hudson/plugins/active_directory/ActiveDirectoryAuthenticationProvider.java [hunk 1]
```diff
@@ -265,6 +268,50 @@ private boolean isAccountDisabled(IADsUser usr) {
...
+    private boolean isAccountExpired(IADsUser usr) {
+        try {
+            Date expirationDate = usr.accountExpirationDate();
+            if (expirationDate != null) {
+                return new Date().after(expirationDate);
+            }
+            return false;
+        } catch (ComException e) {
+            // keep the same behavior as in isAccountDisabled, not sure of the details as the linked page is now a 404
+            if (e.getHRESULT() == 0x8000500D) {
```
Comment (jvz) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectoryAuthenticationProvider.java]: Refactor into constant?

File: src/main/java/hudson/plugins/active_directory/UserAttributesHelper.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+import java.util.Date;
+import java.util.GregorianCalendar;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Ease all the computations required to determine the user account optional attributes for creating
+ * the UserDetails that will be used by the SecurityRealm
+ */
+@Restricted(NoExternalUse.class)
+public class UserAttributesHelper {
```
Comment (jvz) [inline src/main/java/hudson/plugins/active_directory/UserAttributesHelper.java]: Same comments apply as from https://github.com/jenkinsci/ldap-plugin/pull/34

File: src/main/java/hudson/plugins/active_directory/ActiveDirectoryAuthenticationProvider.java (hunk 1 above)
Comment (Wadeck) [reply src/main/java/hudson/plugins/active_directory/ActiveDirectoryAuthenticationProvider.java]: Initially I did not want to touch existing code but as I was using multiple times the value, no problem with that. Thx for the point! I added some documentation on that particular case, that is not a 404.

**Comment by jvz** [comment]
Can we get another review from one of @jglick @stephenc @batmat @rsandell @fbelzunc

**Comment by stephenc** [comment]
🤷‍♀️This is beyond my ken!

**Comment by fbelzunc** [comment]
I will review, merge and release next week. CC @jvz

**Comment by jvz** [comment]
May be related to regression in https://issues.jenkins-ci.org/browse/JENKINS-57607 ?

**Pull request merged**